### PR TITLE
Do not fire InsufficientResourceError when there are intentional reasons

### LINKED ISSUE #45780: DaemonSet FailedPlacement events fire when they shouldn't
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

---

DaemonSet FailedPlacement

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
BUG REPORT
**Kubernetes version** (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.0", GitCommit:"fff5156092b56e6bd60fff75aad4dc9de6b6ef37", GitTreeState:"clean", BuildDate:"2017-03-28T16:36:33Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.3", GitCommit:"0480917b552be33e2dba47386e51decb1a211df6", GitTreeState:"clean", BuildDate:"2017-05-10T15:38:08Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: 
- **OS** (e.g. from /etc/os-release): OL7.3
- **Kernel** (e.g. `uname -a`): 4.1.12-94.2.1.el7uek.x86_64
- **Install tools**:
- **Others**:

**What happened**:

I have a DaemonSet that is only deployed to a subset of nodes in the cluster based on a `nodeSelector`. I'm seeing events fired with reason `FailedPlacement` for the daemonset pods on nodes that do not fit the selector. 

```
Reason: FailedPlacement
Message: failed to place pod on "<node>": Node didn't have enough resource: memory, requested: 9768534016, used: 8361345024, capacity: 14944219136
```

**What you expected to happen**:

These events should never fire because they do not meet the node selectors in the pod spec. 

**How to reproduce it** (as minimally and precisely as possible):

Create a cluster with different sized nodes. E.g. smaller masters than workers. Then deploy the daemonset only to the workers. You will see events triggered for the master nodes that don't meet the requirements. 

**Anything else we need to know**:

### PR DESCRIPTION
**Pull request by gyliu513**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #45780 

**Special notes for your reviewer**:
Return directly of essential predicates failed.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go
Comment (k82cn) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Suggest to avoid order dependency between DaemonSet & Scheduler.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go
Comment (k82cn) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: That'll change the event message of scheduler/kubelet, we should add a release-note for that if it's the solution.

I'm not sure whether we should return immediately when `EssentialPredicates` failed: in predicates, we did not return immediately when any predicate failed although it is enough to skip the node and has a better performance, maybe @wojtek-t / @timothysc can give some input for that.

Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: We are explicitly computing all predicated and avoid early return to give a more actionable feedback for user.
Basically, if you return only first failed predicate, then user may expect that once he fixes this issue it will schedule. But it may not be the case.
If we return e.g. 6 different predicates failing on each node, user expects that it won't be easy to schedule this pod and will need to take a bunch of actions.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: @k82cn - can you clarify this comment? I'm not sure I understand.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Same here - please remove this easly return.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go
Comment (k82cn) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: This PR assume `PodSelectorMatches` will return earlier than `PodFitsResources`, so DaemonSet will not report resource error if node selector is enabled. prefer to avoid such kind of assumption.

Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Oh yeah
+1 for this comment

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go
Comment (gyliu513) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: @wojtek-t The reason that I was trying to return early is that I was trying to ignore the predicate failing caused by `PodFitsResources`, if we get all predicate failures, then the end user will always get the error returned by `PodFitsResources` even if the user already set `nodeSelector`, this is the issue reported by #45780 , any comments for how to avoid such case? Thanks.

Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: It should be handled correctly in daemonset, not in generic scheduler code.

Comment (gyliu513) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Got it, thanks @wojtek-t , then we may need to check the failure reason twice to see if there are intentional reasons first, and do not fire InsufficientResourceError when there are intentional reasons. 

The following is an example of the PR, any comments for this approach?

```
--- a/pkg/controller/daemon/daemoncontroller.go
+++ b/pkg/controller/daemon/daemoncontroller.go
@@ -899,11 +899,35 @@ func (dsc *DaemonSetsController) nodeShouldRunDaemonPod(node *v1.Node, ds *exten
 		return false, false, false, err
 	}

+	// Do not fire InsufficientResourceError when there are intentional reasons
+	fireInsufficientResourceError := true
+	for _, r := range reasons {
+		switch reason := r.(type) {
+		case *predicates.PredicateFailureError:
+			// we try to partition predicates into two partitions here: intentional on the part of the operator and not.
+			switch reason {
+			// intentional
+			case
+				predicates.ErrNodeSelectorNotMatch,
+				predicates.ErrPodNotMatchHostName,
+				predicates.ErrNodeLabelPresenceViolated,
+				// this one is probably intentional since it's a workaround for not having
+				// pod hard anti affinity.
+				predicates.ErrPodNotFitsHostPorts,
+				// DaemonSet is expected to respect taints and tolerations
+				predicates.ErrTaintsTolerationsNotMatch:
+				fireInsufficientResourceError = false
+			}
+		}
+	}
+
 	for _, r := range reasons {
 		glog.V(4).Infof("DaemonSet Predicates failed on node %s for ds '%s/%s' for reason: %v", node.Name, ds.ObjectMeta.Namespace, ds.ObjectMeta.Name, r.GetReason())
 		switch reason := r.(type) {
 		case *predicates.InsufficientResourceError:
-			dsc.eventRecorder.Eventf(ds, v1.EventTypeWarning, FailedPlacementReason, "failed to place pod on %q: %s", node.ObjectMeta.Name, reason.Error())
+			if fireInsufficientResourceError {
+				dsc.eventRecorder.Eventf(ds, v1.EventTypeWarning, FailedPlacementReason, "failed to place pod on %q: %s", node.ObjectMeta.Name, reason.Error())
+			}
 			shouldSchedule = false
 		case *predicates.PredicateFailureError:
 			var emitEvent bool
```

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (gyliu513) [inline pkg/controller/daemon/daemoncontroller_test.go]: I filed an issue here https://github.com/kubernetes/kubernetes/issues/46629 to enable checking the event number

File: pkg/controller/daemon/daemoncontroller.go
Comment (k82cn) [inline pkg/controller/daemon/daemoncontroller.go]: I think it should return immediately: if the node did not match nodeSelector, it should not trigger any event. This PR will still trigger other event, e.g. `ErrNodeUnderMemoryPressure` although nodeSelector did not match, right?

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (k82cn) [inline pkg/controller/daemon/daemoncontroller_test.go]: prefer to combine those two case into one test func, and it's a test cases (func) for nodeSelector.

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (k82cn) [inline pkg/controller/daemon/daemoncontroller_test.go]: Not sure why checking event number, suggest to add `eventRecorder` into `daemonSetsController`, so we can check whether unexpected events are triggered.

BTW, if updating to `syncAndValidateDaemonSets` is required, prefer to handle it in one PR, otherwise, we'll lose the context.

File: pkg/controller/daemon/daemoncontroller.go
Comment (k82cn) [inline pkg/controller/daemon/daemoncontroller.go]: And it's better to extra this logic into a single func, it's too long :(.

Comment (gyliu513) [inline pkg/controller/daemon/daemoncontroller.go]: I think it is old behavior of firing event for `ErrNodeUnderMemoryPressure` even though `nodeSelector` does not match, I'm not touching this logic.

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (gyliu513) [inline pkg/controller/daemon/daemoncontroller_test.go]: `eventRecorder` is already here https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/daemoncontroller_test.go#L257 in `daemonSetsController`, but it is just fake. I did not found a good way for how to check events here, can you elaborate more? Thanks.

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (gyliu513) [inline pkg/controller/daemon/daemoncontroller_test.go]: There are many different cases for node selector, you can check https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/daemoncontroller_test.go#L766-L788 , I was trying to extend the cases a bit by adding one more case for node selector, this case will only cover the node selector. This test case does not related much to this PR, perhaps I can just move this out to a new PR?

**Comment by gyliu513** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by gyliu513** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this
@k8s-bot pull-kubernetes-unit test this

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (k82cn) [inline pkg/controller/daemon/daemoncontroller_test.go]: Add a new member `FakeRecorder` to `daemonSetsController `, so you can get it in this func; After `syncAndValidateDaemonSets`, read from `FakeRecorder.Events` to check what events are triggered.

File: pkg/controller/daemon/daemoncontroller.go
Comment (k82cn) [inline pkg/controller/daemon/daemoncontroller.go]: `InsufficientResourceError` is only an example to me; if nodeSelector did not match, daemonSet should not report other predicates events.

Anyway, I'd like to get other's comments on this :).

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (k82cn) [inline pkg/controller/daemon/daemoncontroller_test.go]: wow, agree, it's better to be a new PR to combine them into one func.

File: pkg/controller/daemon/daemoncontroller.go
Comment (lukaszo) [inline pkg/controller/daemon/daemoncontroller.go]: Please move this to a separate function which just returns true/false. `nodeShouldRunDaemonPod` is already too big.
As for `ErrNodeUnderMemoryPressure` and others I'm actually not sure if DaemonSet will ever see this error. We are using here only `GeneralPredicates` which is not calling `predicates.CheckNodeDiskPressurePredicate`

Comment (gyliu513) [inline pkg/controller/daemon/daemoncontroller.go]: Good point @lukaszo , seems we never hit those `unintentional` predicates errors, shall we propose a new PR to clean those up?

File: pkg/controller/daemon/daemoncontroller.go
Comment (janetkuo) [inline pkg/controller/daemon/daemoncontroller.go]: Why not just return `false, false, false` for intentional `PredicateFailureError`, and move the `InsufficientResourceError` case to after `PredicateFailureError`?

Comment (gyliu513) [inline pkg/controller/daemon/daemoncontroller.go]: Move the `InsufficientResourceError` case to after `PredicateFailureError` may not work if the `Predicates` returns both `InsufficientResourceError` and `ErrNodeSelectorNotMatch` as the `InsufficientResourceError` will be the error in front of `ErrNodeSelectorNotMatch` and the `InsufficientResourceError` will get fired even though we already got `ErrNodeSelectorNotMatch` error.

So here I was first trying to see if there are `intentional` reasons in `shouldFireInsufficientResourceError`, then use its return value to decide if want to fire the `InsufficientResourceError` error.

Also not quite get what does `Why not just return false, false, false for intentional PredicateFailureError` mean? I think we already return `wantToRun, shouldSchedule, shouldContinueRunning = false, false, false` for `intentional` reason?

Comment (janetkuo) [inline pkg/controller/daemon/daemoncontroller.go]: > Move the `InsufficientResourceError` case to after `PredicateFailureError` may not work if the `Predicates` returns both `InsufficientResourceError` and `ErrNodeSelectorNotMatch` as the `InsufficientResourceError` will be the error in front of `ErrNodeSelectorNotMatch` and the `InsufficientResourceError` will get fired even though we already got `ErrNodeSelectorNotMatch` error.

I'm not sure if I understand this part.  If the Predicates returns both `InsufficientResourceError` and `ErrNodeSelectorNotMatch`, we'll just hit the `ErrNodeSelectorNotMatch` and if we return immediately we're done. We won't emit "failed to place pod" event, isn't this what we want?


> Also not quite get what does `Why not just return false, false, false for intentional PredicateFailureError` mean? I think we already return `wantToRun, shouldSchedule, shouldContinueRunning = false, false, false` for `intentional` reason?

Note that we didn't return immediately here, but continue the loop. If we just want to skip emitting "failed to place pod" event when there's an intentional error, why not just return immediately?

Comment (gyliu513) [inline pkg/controller/daemon/daemoncontroller.go]: >I'm not sure if I understand this part. If the Predicates returns both InsufficientResourceError and ErrNodeSelectorNotMatch, we'll just hit the ErrNodeSelectorNotMatch and if we return immediately we're done. We won't emit "failed to place pod" event, isn't this what we want?

We can take a use case as in https://github.com/kubernetes/kubernetes/issues/45780
1) I have two nodes one master and one worker, master node has 10 cpu and 10 memory, worker node has 100 cpu and 100 memory with a label as `foo=bar`.
2) Then I have a demonset with `nodeSelector` as `foo: bar` and also request resources as 20 cpu and 20 memory.
3) After deployment, one pod for the daemonset will be running in worker node, this is what I expected.
4) The problem is that the daemonset event will report it `failed to place pod on "<node>": Node didn't have enough resource: memory, requested: 9768534016, used: 8361345024, capacity: 14944219136` as the master node do not have enough resources, but this is not I want as I have specified the `nodeSelector` for the daemonset.

The fix here is trying to avoid some confusing event, hope this helps ;-)

>Note that we didn't return immediately here, but continue the loop. If we just want to skip emitting "failed to place pod" event when there's an intentional error, why not just return immediately?

You can probably take a look at https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L883-L902 , for master node that I listed above, we will check `noncriticalPredicates` first and then `EssentialPredicates`, the `noncriticalPredicates` will return `InsufficientResourceError` and `EssentialPredicates` will return `ErrNodeSelectorNotMatch`. In the failure reason queue, `InsufficientResourceError` is ahead of `ErrNodeSelectorNotMatch`, so in the switch case here, we will always fire the `InsufficientResourceError` event first. It may not help even if we return directly for `intentional` reasons as this case will be handled after `InsufficientResourceError`.
I want to switch the order of `noncriticalPredicates` and `EssentialPredicates` in https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L883-L902, but it may not be good as we do not want to update the scheduler logic due to daemonset requirement, but should update daemonset to handle such case.

Comment (janetkuo) [inline pkg/controller/daemon/daemoncontroller.go]: Would we want to return error, or emit event when the reason is intentional, such as `ErrNodeSelectorNotMatch`? If not, we check if it's intentional first. If so, return; else do the other check. Something like:

```go 
for _, r := range reasons {
  // only check intentional predicates, if we find any, return false, false, false, nil
}

for _, r := range reasons { 
  switch r.(type) { 
  case *predicates.InsufficientResourceError:
        // emit event, we know we don't have any intentional predicates at this point 
  case *predicates.PredicateFailureError:
       switch reason {
       // no need to check intentional ones 
       case xxx: // unintentional ones 
       case xxx: // unexpected ones 
       default: 
       }
       if emitEvent {xxx}
  }
}
return 
```

Comment (gyliu513) [inline pkg/controller/daemon/daemoncontroller.go]: Yes, the above logic seems more clear, will update the PR soon. One more suggestion is as the function `nodeShouldRunDaemonPod` is already very long, I want to put the check of `intentional predicates` in a new function, comments?

Comment (janetkuo) [inline pkg/controller/daemon/daemoncontroller.go]: Yes, that'd make it cleaner

Comment (gyliu513) [inline pkg/controller/daemon/daemoncontroller.go]: @janetkuo Done

File: pkg/controller/daemon/daemoncontroller.go
Comment (lukaszo) [inline pkg/controller/daemon/daemoncontroller.go]: double `intentional`

Comment (gyliu513) [inline pkg/controller/daemon/daemoncontroller.go]: Done

**Comment by gyliu513** [comment]
@lukaszo @janetkuo I want to make this in 1.7 , can you please show more comments if any? Thanks!

**Comment by k82cn** [comment]
> I want to make this in 1.7 , can you please show more comments if any? Thanks!

And conclusion about this comments https://github.com/kubernetes/kubernetes/pull/46577#discussion_r119509601 ?

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (gyliu513) [inline pkg/controller/daemon/daemoncontroller_test.go]: Done

**Review by k82cn — APPROVED** [review]
lgtm, thanks very much :).

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (gyliu513) [inline pkg/controller/daemon/daemoncontroller_test.go]: I will try to move the `event number` check to `syncAndValidateDaemonSets` in https://github.com/kubernetes/kubernetes/issues/46629 and then if the `event number` is greater than 0, check `event reason` after `syncAndValidateDaemonSets`, will post a PR once this get merged.

File: pkg/controller/daemon/daemoncontroller.go
Comment (gyliu513) [inline pkg/controller/daemon/daemoncontroller.go]: @lukaszo filed an issue https://github.com/kubernetes/kubernetes/issues/46957 here to trace the never hit `unintentional` predicates reasons

File: pkg/controller/daemon/daemoncontroller_test.go
Comment (lukaszo) [inline pkg/controller/daemon/daemoncontroller_test.go]: +1

File: pkg/controller/daemon/daemoncontroller.go
Comment (janetkuo) [inline pkg/controller/daemon/daemoncontroller.go]: daemonset controller didn't fire this error, but emit an event. Also, `InsufficientResourceError` isn't the only case that's skipped. There's typo for "intentional". Update the comment to something like:

```go
// Return directly if there is any intentional predicate failure reason, so that daemonset controller skips
// checking other predicate failures, such as InsufficientResourceError and unintentional errors.
```

File: pkg/controller/daemon/daemoncontroller.go
Comment (janetkuo) [inline pkg/controller/daemon/daemoncontroller.go]: The above comment is less relevant to this function but more relevant to the caller of this function. Update it to:
```go
// hasIntentionalPredicatesReasons checks if any of the given predicate failure reasons 
// is intentional. 
```

**Comment by janetkuo** [comment]
Only nits, LGTM otherwise

File: pkg/controller/daemon/daemoncontroller.go
Comment (gyliu513) [inline pkg/controller/daemon/daemoncontroller.go]: Done

File: pkg/controller/daemon/daemoncontroller.go
Comment (gyliu513) [inline pkg/controller/daemon/daemoncontroller.go]: Done

**Comment by gyliu513** [comment]
@janetkuo and @lukaszo , can you please also make this as milestone 1.7? As only 1.7 can be in submit queue now. Thanks!

**Comment by gyliu513** [comment]
ping @janetkuo and @lukaszo , can you please help check if I can get a /lgtm and /approve label so that this can catch up 1.7? Thanks.

**Review by janetkuo — APPROVED** [review]

File: pkg/controller/daemon/daemoncontroller.go
Comment (mikedanese) [inline pkg/controller/daemon/daemoncontroller.go]: This seems wrong. If a DaemonSet doesn't tolerate NoSchedule but the pod is already running, we need to return false, false, true from nodeShouldRunDaemonPod.

Comment (mikedanese) [inline pkg/controller/daemon/daemoncontroller.go]: Actually it looks like it has been wrong, even before this change.

Comment (luxas) [inline pkg/controller/daemon/daemoncontroller.go]: @mikedanese Is there an open issue for this comment?
Is it a release-blocker to fix this issue?

**Pull request merged**
